File: DaniETH/backend/app/schemas/team.py
# backend/app/schemas/team.py
"""
Schemas Pydantic v2 para Team & Assets.
Solo validación de datos — el almacenamiento es Firestore.
"""
import uuid
from datetime import datetime
from typing import Any

from pydantic import BaseModel, EmailStr, Field


# ── Defaults que se guardan en Firestore ───────────────────────────────────────

DEFAULT_PERMISSIONS: dict[str, bool] = {
    "view_dashboard":    True,
    "run_scans":         False,
    "vulnerability_hub": False,
    "manage_patches":    False,
    "team_management":   False,
    "admin_settings":    False,
}

DEFAULT_NOTIFICATIONS: dict[str, bool] = {
    "email":        True,
    "slack":        True,
    "sms_critical": False,
    "in_app":       True,
}

DEFAULT_NOTIFY_WHEN: dict[str, bool] = {
    "task_assigned":          True,
    "critical_vulnerabilities": True,
    "daily_digest":           False,
}


# ── Team ───────────────────────────────────────────────────────────────────────

class TeamCreate(BaseModel):
    name: str = Field(..., min_length=2, max_length=150)
    description: str | None = None
    icon: str | None = None          # emoji: "🌐", "💻", "🖥️"


class TeamUpdate(BaseModel):
    name: str | None = Field(None, min_length=2, max_length=150)
    description: str | None = None
    icon: str | None = None


class TeamResponse(BaseModel):
    id: str                          # Firestore document ID
    name: str
    description: str | None = None
    icon: str | None = None
    member_count: int = 0
    created_at: str
    updated_at: str


# ── TeamMember ─────────────────────────────────────────────────────────────────

class TeamMemberCreate(BaseModel):
    name: str = Field(..., min_length=2, max_length=255)
    email: EmailStr
    role: str = Field(..., min_length=2, max_length=100)
    team_id: str                     # Firestore doc ID del equipo
    is_team_lead: bool = False
    firebase_uid: str | None = None  # link opcional al usuario de Auth
    permissions: dict[str, bool] | None = None
    notifications: dict[str, bool] | None = None
    notify_when: dict[str, bool] | None = None


class TeamMemberUpdate(BaseModel):
    name: str | None = None
    email: EmailStr | None = None
    role: str | None = None
    team_id: str | None = None
    is_team_lead: bool | None = None
    active_tasks_count: int | None = None
    completed_this_month: int | None = None
    avg_completion_days: float | None = None
    permissions: dict[str, bool] | None = None
    notifications: dict[str, bool] | None = None
    notify_when: dict[str, bool] | None = None


class TeamMemberResponse(BaseModel):
    id: str                          # Firestore document ID
    member_code: str                 # NET-001, APP-002, SRV-003
    firebase_uid: str | None = None
    name: str
    email: str
    role: str
    is_team_lead: bool
    team_id: str
    team_name: str | None = None
    active_tasks_count: int = 0
    completed_this_month: int = 0
    avg_completion_days: float = 0.0
    workload_pct: int = 0
    workload_label: str = "Light Load"
    permissions: dict[str, Any]
    notifications: dict[str, Any]
    notify_when: dict[str, Any]
    created_at: str
    updated_at: str


# ── Asset ──────────────────────────────────────────────────────────────────────

VALID_ASSET_TYPES = {"web_app", "api", "server", "network", "database", "mobile"}
VALID_STATUSES    = {"active", "inactive", "maintenance"}
VALID_ENVS        = {"production", "staging", "development"}


class AssetCreate(BaseModel):
    name: str = Field(..., min_length=2, max_length=255)
    hostname: str = Field(..., min_length=3, max_length=255)
    ip_address: str | None = None
    asset_type: str = "web_app"
    status: str = "active"
    environment: str = "production"
    description: str | None = None
    team_id: str | None = None


class AssetUpdate(BaseModel):
    name: str | None = None
    hostname: str | None = None
    ip_address: str | None = None
    asset_type: str | None = None
    status: str | None = None
    environment: str | None = None
    description: str | None = None
    team_id: str | None = None


class AssetResponse(BaseModel):
    id: str
    name: str
    hostname: str
    ip_address: str | None = None
    asset_type: str
    status: str
    environment: str
    description: str | None = None
    team_id: str | None = None
    team_name: str | None = None
    created_at: str
    updated_at: str


# ── Stats ──────────────────────────────────────────────────────────────────────

class TeamStatsResponse(BaseModel):
    total_members: int
    total_teams: int
    active_tasks: int
    overloaded_count: int    # active_tasks > 5
    available_count: int     # active_tasks < 3
    avg_completion_days: float

File: DaniETH/frontend/src/types/team.ts
// frontend/src/types/team.ts

export interface Team {
  id: string;
  name: string;
  description: string | null;
  icon: string | null;
  member_count: number;
  created_at: string;
  updated_at: string;
}

export interface TeamCreate {
  name: string;
  description?: string;
  icon?: string;
}

export interface TeamUpdate {
  name?: string;
  description?: string;
  icon?: string;
}

// ── Permisos y Notificaciones ──────────────────────────────

export interface MemberPermissions {
  view_dashboard:    boolean;
  run_scans:         boolean;
  vulnerability_hub: boolean;
  manage_patches:    boolean;
  team_management:   boolean;
  admin_settings:    boolean;
}

export interface MemberNotifications {
  email:        boolean;
  slack:        boolean;
  sms_critical: boolean;
  in_app:       boolean;
}

export interface MemberNotifyWhen {
  task_assigned:             boolean;
  critical_vulnerabilities:  boolean;
  daily_digest:              boolean;
}

// ── TeamMember ─────────────────────────────────────────────

export interface TeamMember {
  id: string;
  member_code: string;          // NET-001, APP-002
  firebase_uid: string | null;
  name: string;
  email: string;
  role: string;
  is_team_lead: boolean;
  team_id: string;
  team_name: string | null;
  active_tasks_count: number;
  completed_this_month: number;
  avg_completion_days: number;
  workload_pct: number;         // calculado por el backend
  workload_label: string;       // "Light Load" | "Medium Load" | "Overloaded"
  permissions: MemberPermissions;
  notifications: MemberNotifications;
  notify_when: MemberNotifyWhen;
  created_at: string;
  updated_at: string;
}

export interface TeamMemberCreate {
  name: string;
  email: string;
  role: string;
  team_id: string;
  is_team_lead?: boolean;
  firebase_uid?: string;
  permissions?: Partial<MemberPermissions>;
  notifications?: Partial<MemberNotifications>;
  notify_when?: Partial<MemberNotifyWhen>;
}

export interface TeamMemberUpdate {
  name?: string;
  email?: string;
  role?: string;
  team_id?: string;
  is_team_lead?: boolean;
  active_tasks_count?: number;
  completed_this_month?: number;
  avg_completion_days?: number;
  permissions?: Partial<MemberPermissions>;
  notifications?: Partial<MemberNotifications>;
  notify_when?: Partial<MemberNotifyWhen>;
}

// ── Asset ──────────────────────────────────────────────────

export type AssetType = 'web_app' | 'api' | 'server' | 'network' | 'database' | 'mobile';
export type AssetStatus = 'active' | 'inactive' | 'maintenance';
export type AssetEnvironment = 'production' | 'staging' | 'development';

export interface Asset {
  id: string;
  name: string;
  hostname: string;
  ip_address: string | null;
  asset_type: AssetType;
  status: AssetStatus;
  environment: AssetEnvironment;
  description: string | null;
  team_id: string | null;
  team_name: string | null;
  created_at: string;
  updated_at: string;
}

export interface AssetCreate {
  name: string;
  hostname: string;
  ip_address?: string;
  asset_type?: AssetType;
  status?: AssetStatus;
  environment?: AssetEnvironment;
  description?: string;
  team_id?: string;
}

export interface AssetUpdate {
  name?: string;
  hostname?: string;
  ip_address?: string;
  asset_type?: AssetType;
  status?: AssetStatus;
  environment?: AssetEnvironment;
  description?: string;
  team_id?: string;
}

// ── Stats ──────────────────────────────────────────────────

export interface TeamStats {
  total_members: number;
  total_teams: number;
  active_tasks: number;
  overloaded_count: number;
  available_count: number;
  avg_completion_days: number;
}